UnchainedLauncherCore/src/API/ServerBrowser.cs

```diff
@@ -0,0 +1,248 @@
+            var reqContent = new UpdateServerRequest(info.PlayerCount, info.MaxPlayers, info.CurrentMap);
+            var content = JsonContent.Create(reqContent, options: sOptions);
+            content.Headers.Add("x-chiv2-server-browser-key", key);
+            var httpResponse = await httpc.PutAsync(uri + $"/servers/{info.UniqueId}", content, cs.Token);
+            try
+            {
+                var res = await httpResponse.EnsureSuccessStatusCode()
+                            .Content
+                            .ReadFromJsonAsync<UpdateServerResponse>(sOptions, cs.Token);
+                if (res == null)
+                {
+                    throw new InvalidDataException("Failed to parse response from server");
+                }
+                else
+                {
+                    return res.RefreshBefore;
+                }
+            }
+            catch (InvalidOperationException e)
+            {
+                throw new InvalidDataException("Failed to parse response from server", e);
+            }
+        }
+
+        private static async Task<double> HeartbeatAsync_impl(Uri uri, UniqueServerInfo info, String key)
+        {
+            using CancellationTokenSource cs = new(TimeOutMillis);
+            //var reqContent = new UpdateServerRequest(info.playerCount, info.maxPlayers, info.currentMap);
+            var content = new StringContent("");
+            content.Headers.Add("x-chiv2-server-browser-key", key);
+            var httpResponse = await httpc.PostAsync(uri + $"/servers/{info.UniqueId}/heartbeat", content, cs.Token);
+            try
+            {
+                var res = await httpResponse.EnsureSuccessStatusCode()
+                            .Content
+                            .ReadFromJsonAsync<UpdateServerResponse>(sOptions, cs.Token);
+                if (res == null)
+                {
+                    throw new InvalidDataException("Failed to parse response from server");
+                }
+                else
+                {
+                    return res.RefreshBefore;
+                }
+            }
+            catch (InvalidOperationException e)
+            {
+                throw new InvalidDataException("Failed to parse response from server", e);
+            }
+        }
+
+        private static async Task DeleteServerAsync_impl(Uri uri, UniqueServerInfo info, String key)
+        {
+            using CancellationTokenSource cs = new(TimeOutMillis);
+            //var reqContent = new UpdateServerRequest(info.playerCount, info.maxPlayers, info.currentMap);
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Delete, uri + $"/servers/{info.UniqueId}");
+            request.Headers.Add("x-chiv2-server-browser-key", key);
+            //the DeleteAsync method does not take a content parameter
+            //that's a problem, since that's the only way to set headers
+            //without setting state on the HttpClient
```

Comment:
I think this comment is old.. There's a request setting headers on it above.  Also http spec doesn't support a content body on DELETE IIRC, which is why you can't have it

Reply:
this comment explains why the above code is hand-crafting a delete request and using the general "send this" method instead of simply using the more clear httpc.<method>Async like the other functions do. We don't need (or want) a content body, but we *do* need to set the auth header. For some reason, you need to set the headers *via* the content parameter that DeleteAsync uniquely lacks.

More of an "I know DeleteAsync is a thing that looks like it does precisely this, and here's why I'm not using it" than a "Here's a problem with the present code"

Reply:
Ahh I see